src/ucs/sys/sys.c

```diff
@@ -1615,3 +1616,155 @@ int ucs_sys_is_dynamic_lib(void)
 #endif
 }
 
+typedef struct {
+    double     bw_gbps;       /* Link speed */
+    uint16_t   payload;       /* Payload used to data transfer */
+    uint16_t   tlp_overhead;  /* PHY + data link layer + header + CRC */
+    uint16_t   ctrl_ratio;    /* Number of TLC before ACK */
+    uint16_t   ctrl_overhead; /* Length of control TLP */
+    uint16_t   encoding;      /* Number of encoded symbol bits */
+    uint16_t   decoding;      /* Number of decoded symbol bits */
+    const char *name;         /* Name of PCI generation */
+} ucs_sys_pci_info_t;
+
+/*
+ * - TLP (Transaction Layer Packet) overhead calculations (no ECRC):
+ *   Gen1/2:
+ *     Start   SeqNum   Hdr_64bit   LCRC   End
+ *       1   +   2    +   16      +   4  +  1  = 24
+ *
+ *   Gen3/4:
+ *     Start   SeqNum   Hdr_64bit   LCRC
+ *       4   +   2    +   16      +   4  = 26
+ *
+ * - DLLP (Data Link Layer Packet) overhead calculations:
+ *    - Control packet 8b ACK + 8b flow control
+ *    - ACK/FC ratio: 1 per 4 TLPs
+ *
+ * References:
+ * [1] https://www.xilinx.com/support/documentation/white_papers/wp350.pdf
+ * [2] https://xdevs.com/doc/Standards/PCI/PCI_Express_Base_4.0_Rev0.3_February19-2014.pdf
+ * [3] https://www.nxp.com/docs/en/application-note/AN3935.pdf
+ */
+static const ucs_sys_pci_info_t ucs_sys_pci_info[] = {
+    {.name          = "gen1",
+     .bw_gbps       = 2.5,
+     .payload       = 256,
+     .tlp_overhead  = 24,
+     .ctrl_ratio    = 4,
+     .ctrl_overhead = 16,
+     .encoding      = 8,
+     .decoding      = 10},
+    {.name          = "gen2",
+     .bw_gbps       = 5,
+     .payload       = 256,
+     .tlp_overhead  = 24,
+     .ctrl_ratio    = 4,
+     .ctrl_overhead = 16,
+     .encoding      = 8,
+     .decoding      = 10},
+    {.name          = "gen3",
+     .bw_gbps       = 8,
+     .payload       = 256,
+     .tlp_overhead  = 26,
+     .ctrl_ratio    = 4,
+     .ctrl_overhead = 16,
+     .encoding      = 128,
+     .decoding      = 130},
+    {.name          = "gen4",
+     .bw_gbps       = 16,
+     .payload       = 256,
+     .tlp_overhead  = 26,
+     .ctrl_ratio    = 4,
+     .ctrl_overhead = 16,
+     .encoding      = 128,
+     .decoding      = 130},
+};
+
+ucs_status_t ucs_sys_read_sysfs_file(const char *dev_name,
+                                     const char *sysfs_path,
+                                     const char *file_name, char *output_buffer,
+                                     size_t max, ucs_log_level_t err_level)
+{
+    ssize_t nread;
+
+    if (sysfs_path == NULL) {
+        return UCS_ERR_NO_ELEM;
+    }
+
+    nread = ucs_read_file_str(output_buffer, max, 1, "%s/%s", sysfs_path,
+                              file_name);
+    if (nread < 0) {
+        ucs_log(err_level, "%s: could not read from '%s/%s'", dev_name,
+                sysfs_path, file_name);
+        return UCS_ERR_NO_ELEM;
+    }
+
+    return UCS_OK;
+}
+
+double ucs_sys_get_pci_bw(const char *dev_name, const char *sysfs_path)
```

Comment:
IMO this function (and related helper code) should go to sys/topo.h,c
but ucs_sys_read_sysfs_file should stay in sys/sys.h,c